sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -1189,18 +1189,31 @@ def configure_tx_output_power(self, api, lport, tx_power):
            self.log_error("{} configured tx power {} > maximum power {} supported".format(lport, tx_power, max_p))
         return api.set_tx_power(tx_power)
 
-    def configure_laser_frequency(self, api, lport, freq, grid=75):
-        _, _,  _, lowf, highf = api.get_supported_freq_config()
+    def config_laser_frequency_validate(self, api, lport, freq, grid=75):
+        supported_grid, _,  _, lowf, highf = api.get_supported_freq_config()
         if freq < lowf:
             self.log_error("{} configured freq:{} GHz is lower than the supported freq:{} GHz".format(lport, freq, lowf))
+            return false
         if freq > highf:
             self.log_error("{} configured freq:{} GHz is higher than the supported freq:{} GHz".format(lport, freq, highf))
-        chan = int(round((freq - 193100)/25))
-        if chan % 3 != 0:
-            self.log_error("{} configured freq:{} GHz is NOT in 75GHz grid".format(lport, freq))
+            return false
+        if grid == 75:
+            if (supported_grid >> 7) & 0x1 != 1:
+                self.log_error("{} 75GHz is not supported".format(lport))
+                return False
+            chan = int(round((freq - 193100)/25))
+            if chan % 3 != 0:
+                self.log_error("{} configured freq:{} GHz is NOT in 75GHz grid".format(lport, freq))
+                return False
+        return True
+
+    def configure_laser_frequency(self, api, lport, freq, grid=75):
         if api.get_tuning_in_progress():
             self.log_error("{} Tuning in progress, subport selection may fail!".format(lport))
-        return api.set_laser_freq(freq, grid)
+        try:
```

Comment:
@balram-csco why Xcvrd crash? Do you have the backtrace?

Reply:
If the frequency was not in 75ghz grid  ,routine was not returning failure and asserting  in set_laser_freq causing xcvrd crash. Attaching the crash info below.
CRASH
======
pr 11 12:54:33.881810 sonic NOTICE pmon#xcvrd[470168]: CMIS: Ethernet248: 400G, lanemask=0xff, state=AP_CONFIGURED, appl 1 host_lane_count 8 retries=0
Apr 11 12:54:33.895952 sonic ERR pmon#xcvrd[470168]: CMIS: Ethernet248 configured freq:196000 GHz is NOT in 75GHz grid
Apr 11 12:54:33.905489 sonic ERR pmon#xcvrd[470168]: Exception occured at CmisManagerTask thread due to AssertionError()
Apr 11 12:54:33.906255 sonic ERR pmon#xcvrd[470168]: Traceback (most recent call last):
Apr 11 12:54:33.906284 sonic ERR pmon#xcvrd[470168]:   File "/usr/local/lib/python3.9/dist-packages/xcvrd/xcvrd.py", line 1756, in run
Apr 11 12:54:33.906284 sonic ERR pmon#xcvrd[470168]:     self.task_worker()
Apr 11 12:54:33.906289 sonic ERR pmon#xcvrd[470168]:   File "/usr/local/lib/python3.9/dist-packages/xcvrd/xcvrd.py", line 1656, in task_worker
Apr 11 12:54:33.906316 sonic ERR pmon#xcvrd[470168]:     if 1 != self.configure_laser_frequency(api, lport, freq):
Apr 11 12:54:33.906316 sonic ERR pmon#xcvrd[470168]:   File "/usr/local/lib/python3.9/dist-packages/xcvrd/xcvrd.py", line 1376, in configure_laser_frequency
Apr 11 12:54:33.906322 sonic ERR pmon#xcvrd[470168]:     return api.set_laser_freq(freq, grid)
Apr 11 12:54:33.906322 sonic ERR pmon#xcvrd[470168]:   File "/usr/local/lib/python3.9/dist-packages/sonic_platform_base/sonic_xcvr/api/public/c_cmis.py", line 189, in set_laser_freq
Apr 11 12:54:33.906329 sonic ERR pmon#xcvrd[470168]:     assert channel_number % 3 == 0
Apr 11 12:54:33.906329 sonic ERR pmon#xcvrd[470168]: AssertionError
Apr 11 12:54:33.906433 sonic ERR pmon#xcvrd[470168]: Xcvrd: exception found at child thread CmisManagerTask due to AssertionError()
Apr 11 12:54:33.906445 sonic ERR pmon#xcvrd[470168]: Exiting main loop as child thread raised exception!
Apr 11 12:54:33.925921 sonic INFO pmon#supervisord 2024-04-11 12:54:33,925 INFO exited: xcvrd (terminated by SIGKILL; not expected)
Apr 11 12:54:34.927600 sonic INFO pmon#supervisord 2024-04-11 12:54:34,927 INFO spawned: 'xcvrd' with pid 470209
Apr 11 12:54:35.048930 sonic NOTICE pmon#xcvrd[470209]: Starting up...
Apr 11 12:54:35.048930 sonic NOTICE pmon#xcvrd[470209]: XCVRD INIT: Start daemon init...
Apr 11 12:54:38.733590 sonic INFO pmon#supervisord: xcvrd DBG _optics_init_once:OPTICS_INIT_ONCE: start one time optics lib initialization
Apr 11 12:54:38.733677 sonic INFO pmon#supervisord: xcvrd Debug thread created!
Apr 11 12:54:38.733677 sonic INFO pmon#supervisord: xcvrd ERR optics_lib_shm_init:SHM_INIT: Shared memory exists, treat as process restart case
Apr 11 12:54:38.733677 sonic INFO pmon#supervisord: xcvrd DBG optics_lib_shm_init:SHM_DONE: restart_count 12
Apr 11 12:54:38.733677 sonic INFO pmon#supervisord: xcvrd DBG _optics_init_once:OPTICS_INIT_ONCE: one time opt